lib/AST/ASTDumper.cpp

```diff
@@ -114,27 +135,62 @@ namespace {
     }
 
   };
-} // end anonymous namespace
 
-static void printGenericParameters(raw_ostream &OS, GenericParamList *Params) {
-  if (!Params)
-    return;
-  OS << ' ';
-  Params->print(OS);
-}
+/// Wraps a \c raw_ostream so that anything printed through it is automatically
+/// escaped appropriately for a double-quoted string.
+class escaping_ostream : public raw_ostream {
+  raw_ostream &base_os;
 
-static void printSourceRange(raw_ostream &OS, const SourceRange R,
-                             const ASTContext &Ctx) {
-  if (!R.isValid())
-    return;
+public:
+  escaping_ostream(raw_ostream &base_os)
+    : raw_ostream(/*unbuffered=*/true), base_os(base_os)
+  {}
 
-  PrintWithColorRAII(OS, RangeColor) << " range=";
-  R.print(PrintWithColorRAII(OS, RangeColor).getOS(), Ctx.SourceMgr,
-          /*PrintText=*/false);
-}
+  virtual ~escaping_ostream() {}
+
+  virtual void reserveExtraSpace(uint64_t ExtraSize) override {
+    base_os.reserveExtraSpace(ExtraSize);
+  }
+
+  virtual raw_ostream &changeColor(enum Colors Color, bool Bold = false,
+                                   bool BG = false) override {
+    return base_os.changeColor(Color, Bold, BG);
+  }
+
+  virtual raw_ostream &resetColor() override {
+    return base_os.resetColor();
+  }
+
+  virtual raw_ostream &reverseColor() override {
+    return base_os.reverseColor();
+  }
+
+  virtual bool is_displayed() const override {
+    return base_os.is_displayed();
+  }
+
+  virtual bool has_colors() const override {
+    return base_os.has_colors();
+  }
+
+  virtual void enable_colors(bool enable) override {
+    base_os.enable_colors(enable);
+  }
+
+private:
+  virtual void write_impl(const char *Ptr, size_t Size) override {
+    base_os.write_escaped(StringRef(Ptr, Size), /*UseHexEscapes=*/true);
```

Comment:
FYI, this made the tau in generic parameter names print as escape sequences:
```
(primary_archetype_type address=true name="\xCF\x84_0_0"
  (interface_type=generic_type_param_type depth=0 index=0))
```

Reply:
True. Using `write_escaped` happened to be very convenient, but I wouldn't be averse to a patch that excepted the tau character, or some larger set of known-printable characters (as long as it'll still escape various newlines, quotes, control characters, etc.).